File: tests/wasmbackend/tglobals.js
// Driver for tglobals.nim: call bump() repeatedly and observe the module-level
// globals persist across calls in the shared linear memory.
"use strict";
const fs = require("fs");
const bytes = fs.readFileSync(process.argv[2]);
const mod = new WebAssembly.Module(bytes);
const memory = new WebAssembly.Memory({ initial: 1 });
const env = { memory };
for (const imp of WebAssembly.Module.imports(mod)) {
  if (imp.module === "env" && imp.kind === "function" && !(imp.name in env)) {
    env[imp.name] = (a) => a;
  }
}
const e = new WebAssembly.Instance(mod, { env }).exports;
console.log("bump() =", e.bump());
console.log("bump() =", e.bump());
console.log("bump() =", e.bump());
console.log("getTotal() =", e.getTotal());

File: tests/wasmbackend/tglobals.nim
# Module-level globals as fixed memory slots (zero-initialised — linear memory
# starts at 0). State persists across exported calls in the shared buffer, just
# as it does in the JS backend's `mem`.

var counter: int32
var total: int32

proc bump(): int32 {.exportc.} =
  counter = counter + 1'i32
  total = total + counter
  result = counter

proc getTotal(): int32 {.exportc.} =
  result = total
